frame/staking/src/tests.rs

```diff
@@ -3827,88 +3827,103 @@ fn on_finalize_weight_is_nonzero() {
-
-		// verify preconditions
-		let weight_of = Staking::weight_of_fn();
-		let node = Node::<Test>::from_id(&stash).unwrap();
-		assert_eq!(
-			{
-				let origin_bag = Bag::<Test>::get(node.bag_upper).unwrap();
-				origin_bag.iter().count()
-			},
-			1,
-			"stash should be the only node in origin bag",
-		);
-		let other_node = Node::<Test>::from_id(&other_stash).unwrap();
-		assert!(!other_node.is_misplaced(&weight_of), "other stash balance never changed");
-		assert_ne!(
-			{
-				let destination_bag = Bag::<Test>::get(other_node.bag_upper);
-				destination_bag.iter().count()
-			},
-			0,
-			"destination bag should not be empty",
-		);
+	#[test]
+	fn rebag_works() {
+		ExtBuilder::default().build_and_execute(|| {
+			// add a nominator to genesis state
+			bond_nominator(42, 43, 20, vec![11]);
+			Balances::make_free_balance_be(&42, 2_000);
 
-		// Update `stash`'s value to match `other_stash`, and bond extra to update its weight.
-		//
-		// This implicitly calls rebag, so the user stays in the best bag they qualify for.
-		let new_balance = <Test as Config>::Currency::free_balance(&other_stash);
-		<Test as Config>::Currency::make_free_balance_be(&stash, new_balance);
-		Staking::bond_extra(
-			RawOrigin::Signed(stash.clone()).into(),
-			new_balance,
-		).unwrap();
-
-		// node should no longer be misplaced
-		// note that we refresh the node, in case the storage value has changed
-		let node = Node::<Test>::from_id(&stash).unwrap();
-		assert!(!node.is_misplaced(&weight_of), "node must be in proper place after rebag");
-	});
+			// given
+			assert_eq!(get_bags(), vec![(10, vec![31]), (20, vec![42]), (1000, vec![11, 21, 101])]);
+
+			// increase stake and implicitly rebag with `bond_extra` to the level of non-existent bag
+			assert_ok!(Staking::bond_extra(Origin::signed(42), 1_980)); // 20 + 1_980 = 2_000
+			assert_eq!(
+				get_bags(),
+				vec![(10, vec![31]), (20, vec![]), (1000, vec![11, 21, 101]), (2000, vec![42])]
+			);
+
+			// decrease stake within the range of the current bag
+			assert_ok!(Staking::unbond(Origin::signed(43), 999)); // 2000 - 999 = 1001
+			assert_ok!(Staking::rebag(Origin::signed(43), 42));
+			// does not change bags
+			assert_eq!(
+				get_bags(),
+				vec![(10, vec![31]), (20, vec![]), (1000, vec![11, 21, 101]), (2000, vec![42])]
+			);
+
+			// reduce stake to the level of a non-existent bag
+			assert_ok!(Staking::unbond(Origin::signed(43), 971)); // 1001 - 971 = 30
+			assert_ok!(Staking::rebag(Origin::signed(43), 42));
+			// creates the bag and moves the voter into it
+			assert_eq!(
+				get_bags(),
+				vec![
+					(10, vec![31]),
+					(20, vec![]),
+					(30, vec![42]),
+					(1000, vec![11, 21, 101]),
+					(2000, vec![])
+				]
+			);
+
+			// increase stake by `rebond`-ing to the level of a pre-existing bag
+			assert_ok!(Staking::rebond(Origin::signed(43), 31)); // 30 + 41 = 61
+			assert_ok!(Staking::rebag(Origin::signed(43), 42));
```

Comment:
again, should these not call rebag automatically?

Reply:
fixed it in [`e7fdad2` (#9422)](https://github.com/paritytech/substrate/pull/9422/commits/e7fdad2aa55c5e71ee209a11baf8b3014248f260)

Reply:
887035169fabb7778e806c8a6f36fa6c092fe5e2